zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/identity/IdentitySetupInitializeProcessor.java

```diff
@@ -40,261 +29,61 @@
-                        },
-                        () -> createUser(user, role, tenant)));
+    final var defaultRole = setupRecord.getDefaultRole();
+    final var roleExists = roleState.getRole(defaultRole.getRoleId()).isPresent();
+    createDefaultRole(initializationKey, defaultRole);
 
-    record.getMappings().stream()
-        .map(MappingRecord.class::cast)
-        .forEach(
-            mapping ->
-                mappingState
-                    .get(mapping.getMappingId())
-                    .ifPresentOrElse(
-                        persistedMapping -> {
-                          assignEntityToRole(
-                              role, persistedMapping.getMappingId(), EntityType.MAPPING);
-                          assignEntityToTenant(tenant, mapping.getMappingId(), EntityType.MAPPING);
-                        },
-                        () -> createMapping(mapping, role, tenant)));
-  }
-
-  private void createRole(final RoleRecord role) {
-    stateWriter.appendFollowUpEvent(role.getRoleKey(), RoleIntent.CREATED, role);
-  }
+    final var defaultTenant = setupRecord.getDefaultTenant();
+    createDefaultTenant(initializationKey, defaultTenant);
 
-  private void createUser(final UserRecord user, final RoleRecord role, final TenantRecord tenant) {
-    stateWriter.appendFollowUpEvent(user.getUserKey(), UserIntent.CREATED, user);
-    assignEntityToRole(role, user.getUsername(), EntityType.USER);
-    assignEntityToTenant(tenant, user.getUsername(), EntityType.USER);
-  }
+    final var users = setupRecord.getUsers();
+    createUsers(initializationKey, users, defaultRole, defaultTenant);
 
-  private void createTenant(final TenantRecord tenant) {
-    stateWriter.appendFollowUpEvent(tenant.getTenantKey(), TenantIntent.CREATED, tenant);
-  }
-
-  private void createMapping(
-      final MappingRecord mapping, final RoleRecord role, final TenantRecord tenant) {
-    stateWriter.appendFollowUpEvent(mapping.getMappingKey(), MappingIntent.CREATED, mapping);
-    assignEntityToRole(role, mapping.getMappingId(), EntityType.MAPPING);
-    assignEntityToTenant(tenant, mapping.getMappingId(), EntityType.MAPPING);
-  }
+    final var mappings = setupRecord.getMappings();
+    createMappings(initializationKey, mappings, defaultRole, defaultTenant);
 
-  private boolean assignEntityToRole(
-      final RoleRecord role, final String entityId, final EntityType entityType) {
-    final var roleId = role.getRoleId();
-    final var roleKey = role.getRoleKey();
-    final var isAlreadyAssigned =
-        switch (entityType) {
-          case USER, MAPPING ->
-              membershipState.hasRelation(entityType, entityId, RelationType.ROLE, roleId);
-          default -> throw new IllegalArgumentException("Unsupported entity type: " + entityType);
-        };
-    if (isAlreadyAssigned) {
-      return false;
+    if (!roleExists) {
+      addAllPermissions(initializationKey, defaultRole);
     }
```

Comment:
ℹ️ I purposely put the permissions at the end of the processor. I initially had it grouped together with the role, but by having it at the end it gave me an easy deterministic way of knowing when the initialization is complete in our tests. A bit hacky, but it works for now and I expect there's going to be some changes to this processor in the near-future anyway.